src/client/Microsoft.Identity.Client/ManagedIdentity/CloudShellManagedIdentitySource.cs

```diff
@@ -0,0 +1,67 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Globalization;
+using System.Net.Http;
+using Microsoft.Identity.Client.Core;
+using Microsoft.Identity.Client.Internal;
+
+namespace Microsoft.Identity.Client.ManagedIdentity
+{
+    /// <summary>
+    /// Original source of code: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/identity/Azure.Identity/src/CloudShellManagedIdentitySource.cs
+    /// </summary>
+    internal class CloudShellManagedIdentitySource : ManagedIdentitySource
+    {
+        private readonly Uri _endpoint;
+
+        public static ManagedIdentitySource TryCreate(RequestContext requestContext)
+        {
+            string msiEndpoint = EnvironmentVariables.MsiEndpoint;
+
+            // if ONLY the env var MSI_ENDPOINT is set the MsiType is CloudShell
+            if (string.IsNullOrEmpty(msiEndpoint))
+            {
+                requestContext.Logger.Verbose("[Managed Identity] Cloud shell managed identity is unavailable.");
+                return null;
+            }
+
+            Uri endpointUri;
+            try
+            {
+                endpointUri = new Uri(msiEndpoint);
+            }
+            catch (FormatException ex)
+            {
+                requestContext.Logger.Error("[Managed Identity] Invalid endpoint found for the environment variable MSI_ENDPOINT: " + msiEndpoint);
+                throw new MsalClientException(MsalError.InvalidManagedIdentityEndpoint, string.Format(
+                    CultureInfo.InvariantCulture, MsalErrorMessage.ManagedIdentityEndpointInvalidUriError, "MSI_ENDPOINT", msiEndpoint, "Cloud Shell"), ex);
+            }
+
+            requestContext.Logger.Verbose("[Managed Identity] Creating cloud shell managed identity. Endpoint URI: " + msiEndpoint);
+            return new CloudShellManagedIdentitySource(endpointUri, requestContext);
+        }
+
+        private CloudShellManagedIdentitySource(Uri endpoint, RequestContext requestContext) : base(requestContext)
+        {
+            _endpoint = endpoint;
+            if (!string.IsNullOrEmpty(requestContext.ServiceBundle.Config.ManagedIdentityUserAssignedClientId) || null != requestContext.ServiceBundle.Config.ManagedIdentityUserAssignedResourceId)
```

Comment:
Thoughts on refactoring these two properties? Combine them into one, something like `UserAssignedManagedIdentityId` and when it's added to the request, if it's a guild add it as client ID, or resource ID. We don't really need to differentiate anywhere else in the code.

Reply:
Updated.